Help: Update contact options on upgrade

jeremeylduvall opened the pull request:
This is an Intended fix for #1827 by updating the contact form details when the user cancels or purchases an upgrade. 

This PR does a few different things:

1. I added a new action type called `OLARK_NOT_READY` in `/lib/olark-store/index.js` that sets `isOlarkReady` to `false`.
2. I added an action in `lib/olark-store/actions.js` that calls the dispatcher referencing the action type `OLARK_NOT_READY` mentioned above.
3. The action is then called when a user cancels or removes a purchase as well as when we load the Thank you page after a purchase.
4. In `lib/olark/index.js`, I created a new function called `fetchUserEligibility`, which basically just gets the correct configuration based on the user information
5. Finally, in `me/help/help-contact/index.jsx`, I added a call to `fetchUserEligibility` in `componentWillMount`. It checks whether `isOlarkReady` is set to `false,` which now happens when users purchase or cancel an upgrade. If it's set to `false,` it re-grabs the correct configuration.

The branch currently works correctly (screenshots below), but I have a few follow-up questions:

1. Is there a better place to put the action for users cancelling/removing/adding a purchase? Dropping the call in `cancelPurchase` and `removePurchase` seems fine, but I couldn't find a great spot for when completing a purchase.
2. ESLint through an error on `client/me/help/help-contact/index.jsx` because we define `olark` both up top and below. I tried exporting just the function `fetchUserEligibility` from `lib/olark/index.js`, but I couldn't get that to work. Is there a better way to export just the method so I don't have to import `olark` at the top and re-use that name in `help-contact/index.jsx`?

**To test:**
1. Load up this branch and login using an account without any upgrades
2. Visit https://wordpress.com/help/contact. You should see a forum contact form.
3. Without reloading, purchase an upgrade by clicking "Manage Purchases"
4. With the upgrade purchased, navigate back to the contact page without reloading (click "Me" and then "Help" on the left-hand side). You should now see a contact form/live chat box.
5. Click "Manage Purchases" again and cancel the upgrade.
6. Navigate back to the contact page without reloading. You should see a forum submission box.

Before:
![screen shot 2016-04-11 at 9 56 36 am](https://cloud.githubusercontent.com/assets/7240478/14433587/170d114e-ffcc-11e5-81b8-360c63241359.png)

After purchase (no reload):
![screen shot 2016-04-11 at 9 57 05 am](https://cloud.githubusercontent.com/assets/7240478/14433592/1ddeb068-ffcc-11e5-878e-1046e7253ffd.png)

omarjackman commented on client/lib/olark/index.js:
```diff
@@ -151,6 +151,12 @@ const olark = {
 
+	fetchUserEligibility() {
+		this.getOlarkConfiguration()
+			.then( ( configuration ) => this.configureOlark( configuration ) )
```
Calling `configureOlark` more than once in the page lifetime will cause unpredictable issues because we'll end up listening to for [these](https://github.com/Automattic/wp-calypso/blob/fix/1827-update-contact-option/client/lib/olark/index.js#L118) events more than once.

omarjackman commented on client/lib/olark/index.js:
```diff
@@ -151,6 +151,12 @@ const olark = {
 		}
 	},
 
+	fetchUserEligibility() {
```
Calling `configureOlark` more than once in the page lifetime will cause unpredictable issues because we'll end up listening to [these](https://github.com/Automattic/wp-calypso/blob/fix/1827-update-contact-option/client/lib/olark/index.js#L118) events more than once.

I think what you should do here is create a new function based on [this](https://github.com/Automattic/wp-calypso/blob/fix/1827-update-contact-option/client/lib/olark/index.js#L173) method that uses the Olark Api to change the configuration.

Your new method could look something like this:
```
	fetchUserEligibility() {
		this.getOlarkConfiguration()
			.then( ( configuration ) => this.updateOlarkGroupAndEligibility( configuration ) )
			.catch( ( error ) => this.handleError( error ) );
	},
	updateOlarkGroupAndEligibility( wpcomOlarkConfig = {} ) {
		const isUserEligible = ( 'undefined' === typeof wpcomOlarkConfig.isUserEligible ) ? true : wpcomOlarkConfig.isUserEligible;
		olarkApi( 'api.chat.setOperatorGroup', { group: wpcomOlarkConfig.group } );
		olarkActions.setUserEligibility( isUserEligible );
	}
```
See [api.chat.setOperatorGroup](https://www.olark.com/api#api.chat.setOperatorGroup)

omarjackman replied:
After thinking about this more maybe `fetchUserEligibility` should be named something else since it actually fetches the olark configuration and then updates the olark group and user eligibility. Nothing comes to mind right now

omarjackman commented on client/lib/upgrades/actions/purchases.js:
```diff
@@ -19,6 +20,8 @@ const PURCHASES_FETCH_ERROR_MESSAGE = i18n.translate( 'There was an error retrie
 	PURCHASE_REMOVE_ERROR_MESSAGE = i18n.translate( 'There was an error removing the purchase.' );
 
 function cancelPurchase( purchaseId, onComplete ) {
+	olarkActions.setNotReady();
```
I think we could eliminate the `OLARK_NOT_READY` and `setNotReady` actiontype/action and just call the new [`olark.fetchUserEligibility()`](https://github.com/Automattic/wp-calypso/pull/4649/files#r59305069) here.

omarjackman commented on client/lib/upgrades/actions/purchases.js:
```diff
@@ -163,6 +166,8 @@ function fetchUserPurchases( userId ) {
 }
 
 function removePurchase( purchaseId, userId, onComplete ) {
+	olarkActions.setNotReady();
```
[And](https://github.com/Automattic/wp-calypso/pull/4649/files#r59305593) here.

omarjackman commented on client/me/help/help-contact/index.jsx:
```diff
@@ -51,6 +52,12 @@ module.exports = React.createClass( {
 		olarkActions.hideBox();
 	},
 
+	componentWillMount: function() {
```
And we wouldn't need this

omarjackman commented on client/my-sites/upgrades/controller.jsx:
```diff
@@ -208,7 +208,10 @@ module.exports = {
+			receiptId = Number( context.params.receiptId ),
+			olarkActions = require( 'lib/olark-store/actions' );
+
+		olarkActions.setNotReady();
```
Or this

omarjackman commented:
I didn't see any errors but I did notice that calling `fetchUserEligibility` as you've written it would cause olark events to end up with double listeners installed. I offered an alternative approach that would just call the necessary methods that you were looking for.

jeremeylduvall commented:
Well that's much easier. Thanks @omarjackman! I updated a few things based on your recommendations and removed the action calls. Now, `client/lib/olark/index.js` has two new functions—`fetchOlarkConfigUpdate` and `updateOlarkGroupAndEligibility`. 

I modified the calls that are triggered on upgrade removal/addition a bit. I'm looking at three aspects:

1. Adds a new purchase
2. Cancels a purchase
3. Removes a purchase

We no longer call `fetchOlarkConfigUpdate` when we load the Thank You page. Instead, it's being called in `clearPurchases` in `lib/upgrades/actions/purchases.js`. We call `clearPurchases` both [when a purchase is removed](https://github.com/Automattic/wp-calypso/blob/f70ae76e3740c1e262e9be064669615641748491/client/lib/upgrades/actions/purchases.js#L28) and [when a new purchase is made](https://github.com/Automattic/wp-calypso/blob/214626a9db99021fd50dceddd589cf4c4416d87c/client/my-sites/upgrades/checkout/checkout.jsx#L146). So 1 and 2 are taken care of. For 3, I added a call [in the `else` clause here](https://github.com/Automattic/wp-calypso/blob/f70ae76e3740c1e262e9be064669615641748491/client/lib/upgrades/actions/purchases.js#L177) so we only call the update if the removal was successful.

Still works as originally anticipated.

omarjackman commented:
@jeremeylduvall I like these changes though I think the name `fetchOlarkConfigUpdate` could be more descriptive to indicate that its actually only updating the olark group and eligibility. Maybe we could combine `fetchOlarkConfigUpdate` and `updateOlarkGroupAndEligibility` to make it easier. When thats considered feel free to ship.

LGTM 🚢

jeremeylduvall commented:
Thanks @omarjackman. I combined the functions so there's just `updateOlarkGroupAndEligibility` getting added now. If you could spot check one more time before I ship, it would be much appreciated!

omarjackman commented:
Looks good :shipit:
